### PR TITLE
Add customizable URL-dispatching

### LINKED ISSUE #817: Overhaul URL dispatcher (magnet support, customizable)
Now that our request-resource-hook is almighty, we can move the URL dispatch from `parse-url` to `request-resource-hook`.

Ideally, we should make it in a way that's easily customizable.
For instance, we need to derive a way to let the user customize which Torrent client to use to open magnet links.
Maybe there is nothing to do beside defining a handler that checks which Torrent client is available and loads the magnet link appropriately.

In the end, the following expression can be removed from `parse-url`:
```lisp
((str:starts-with? "magnet:" input-url)
             (log:debug "Open magnet link with external application.")
             (ignore-errors
              (uiop:launch-program (list "xdg-open" input-url))
              (nyxt/minibuffer-mode:cancel-input)))
```

### PR DESCRIPTION
**Pull request by aartaka**
This adds a mechanism for a customizable URL-dispatching and a simple way to customize it.

### Motivation

There can be a necessity to change browser behavior while encountering a certain type of link (e.g., open torrent client on "magnet:" link, redirect "doi:" protocol link to the DOI website). Using handlers and `request-resource-hook` is too much of a work for a simple task of doing something depending on domain/scheme/host/whatever in a URL. There's a need in a simpler way to act on URLs while having robustness of hooks.

### What's Not There (Yet)

- Shorter syntax for shell command invocations on matching URLs. Although it can be built on top of this URL dispatching mechanism easily, it needs more discussion.

### How Has This Been Tested

Browser was started with this config file:
``` lisp
(in-package #:nyxt-user)

(define-configuration browser
  ((url-dispatchers `(,(url-dispatching-handler
                        (match-scheme "doi")
                        (lambda (url)
                          (quri:uri (format nil "https://doi.org/~a"
                                            (quri:uri-path url))))
                        "Resolve DOI links")
                      ,(url-dispatching-handler
                        (match-host "reddit.com")
                        (lambda (url)
                          (quri:copy-uri url :host "old.reddit.com"))
                        "Open old Reddit")
                      ,(url-dispatching-handler
                        (match-scheme "javascript")
                        (lambda (url)
                          (ffi-buffer-evaluate-javascript
                           (current-buffer) (quri:uri-path url))
                          nil))
                      ,@%slot-default))))
```
And the given handlers worked in the appropriate situations. No bugs were noticed.

I'll be glad to see your feedback on that!

Closes #817

### CODE REVIEW

**Comment by Ambrevar** [comment]
Thanks, it's a great start.

This raises a design question though. Now I understand what you intention was in your last comment, but I don't understand the reason: why create a new hook?
Why not create a "handler contruction helper"?
Something like that (untested):

```lisp
(define-configuration browser
    ((request-resource-hook
      (reduce #'hooks:add-hook
              (url-dispatching-handler
               (match-scheme "doi")
               (lambda (url)
                 (quri:uri (format nil "https://doi.org/~a"
                                   (quri:uri-path url))))
               :initial-value %slot-default)))))
```

**Comment by Ambrevar** [comment]
(The `url-dispatching-handler` helpder function would then return a resource handler.)

**Comment by aartaka** [comment]
There are two weak reasons:
- User convenience: it's easier to make a list of simple handlers and to leave composing them to some other function.
- Narrow focus: if this function is focused on the URL transformation and redirection only, then it should be `uri->uri` handler. If it will be `resource` handler, it'll lose this focus and idea. Something can go wrong because of that, either conceptually (user intentionally would use it to modify `request-data`) or implementationally (it will be implemented in a way that would be leaky and allow for these same `request-data` side-effects).

Again, these reasons are weak and rewriting it for `request-resource-hook` sounds better to me.

However, I don't have a clear idea of how it could look (at least yet). Your example is valid, but having 3+ dispatching handlers and some other `resource` handlers would be hard to read. You can prove me wrong though.

Maybe add a dedicated function to abstract this `(reduce #'hooks:add-hook ... )` or am I trying to abstract too much? :)

**Comment by Ambrevar** [comment]
Some counter-points:

- The `(reduce ...)` only adds one line.
- The `url-dispatching-handler` helper is the center of the abstraction here.  It abstracts away the request-data and leaves the user only with the quri:uri.
- We can store the list of URI dispatches in a separate list.  It's a full-blown programming language after all! :)  Example


```lisp
(defvar my-dispatchers
  (list
   (url-dispatching-handler
    (match-scheme "doi")
    (lambda (url)
      (quri:uri (format nil "https://doi.org/~a"
                        (quri:uri-path url)))))
   (url-dispatching-handler
    (match-scheme "magnet")
    (lambda (url)
      ...))))

(define-configuration browser
    ((request-resource-hook
      (reduce #'hooks:add-hook
              (append
               my-dispatchers
               my-non-dispatchers)
              :initial-value %slot-default))))
```

Do you like it?

**Comment by aartaka** [comment]
Yes, you're right. I've pushed the changes you suggested. 

I've added a `hyphenize-dispatching-test` function -- it can seem unnecessary now, but it can be handy for smart per domain mode enabling/disabling from #376. Specifically, it can allow us to distinguish the hosts/domains that either the browser or the user have set rules for -- we can search them by names!

**Comment by Ambrevar** [comment]
Sorry, I forgot that our handlers had to be named :)

Instead of the macro and gensym, what about adding a `:name` key argument to the  `url-dispatching-handler` helper?

**Comment by Ambrevar** [comment]
Then we would not need a macro nor `hyphenize-dispatching-test`.

File: source/browser.lisp [hunk 1]
```diff
@@ -1102,6 +1102,69 @@ Deal with REQUEST-DATA with the following rules:
...
+
+Example:
+\(define-configuration buffer
+   ((request-resource-hook (hooks:add-hook
+                            (url-dispatching-handler
+                                (match-scheme \"doi\")
+                                (lambda (url)
+                                  (quri:uri (format nil \"https://doi.org/~a\"
+                                                    (quri:uri-path url))))
+                                \"Resolve DOI links\")
```
Comment (Ambrevar) [inline source/browser.lisp:1134]: I'd add the magnet example since it shows how to open URI with external programs.

Comment (aartaka) [reply source/browser.lisp:1134]: Sure.

**Comment by aartaka** [comment]
> Instead of the macro and gensym, what about adding a `:name` key argument to the `url-dispatching-handler` helper?

Yes, I thought about it myself. I'll push it in an hour!

File: source/browser.lisp [hunk 2]
```diff
@@ -1102,6 +1102,66 @@ Deal with REQUEST-DATA with the following rules:
          (log:debug "Forwarding: ~a" (object-display url))
          request-data)))))
 
+(export-always 'url-dispatching-handler)
+(defun url-dispatching-handler (test action &optional (description "") (name (gensym)))
```
Comment (Ambrevar) [inline source/browser.lisp:1106]: The philosophy of our hook system is to be legible, unlike Emacs hooks which accept lambda.
`gensym` hides the intention away from the user.  This is why I suggest that we do like the rest of the Nyxt hook system and we force the use of a name.

Actually, if names are mandatory, aren't the description and the name redundant?  If so, then feel free to remove `description`.

File: source/browser.lisp [hunk 3]
```diff
@@ -1102,6 +1102,66 @@ Deal with REQUEST-DATA with the following rules:
...
+  "Returns a `resource' handler that, if `add-hook'ed to the `request-resource-hook',
```
Comment (Ambrevar) [inline source/browser.lisp:1107]: Use the imperative form `Return`.

File: source/browser.lisp [hunk 4]
```diff
@@ -1102,6 +1102,66 @@ Deal with REQUEST-DATA with the following rules:
...
+will automatically apply its ACTION on the URLs that conform to TEST.
+
+TEST and ACTION should be the functions of one argument -- the requested URL.
```
Comment (Ambrevar) [inline source/browser.lisp:1110]: Typo: "should be functions".
Nit: I would replace the `--` by a comma.  As you prefer.

File: source/browser.lisp [hunk 5]
```diff
@@ -1102,6 +1102,66 @@ Deal with REQUEST-DATA with the following rules:
...
+In case ACTION returns nil, URL request is aborted.
+New URL returned by ACTION is loaded otherwise.
```
Comment (Ambrevar) [inline source/browser.lisp:1112]: The new URL

File: source/browser.lisp [hunk 6]
```diff
@@ -1102,6 +1102,66 @@ Deal with REQUEST-DATA with the following rules:
...
+
+`match-host', `match-scheme', `match-domain' and `match-file-extension'
+can be used to create TEST-functions.
```
Comment (Ambrevar) [inline source/browser.lisp:1115]: This sounds like they are the only valid test functions.  I would rephrase this and list them as example functions.

**Comment by Ambrevar** [comment]
I guess we are almost good to merge (sorry for being picky :p).

Previously we discussed the convention that the `URI` symbol would refer to `QURI:URI` objects, while `URL` would refer to strings.  Do you like this convention?  If so, please adapt your code to follow it.  Thanks! :)

**Comment by aartaka** [comment]
> I guess we are almost good to merge (sorry for being picky :p).

It makes for good code, so it's nothing to be sorry about :)

> Previously we discussed the convention that the `URI` symbol would refer to `QURI:URI` objects, while `URL` would refer to strings. Do you like this convention? If so, please adapt your code to follow it. Thanks! :)

Hmmmm, I'm not sure there. Several reasons to doubt this distinction:
- Terms URI and URL are ill-defined and are frequently used interchangeably, so distinguishing between these is hard.
- Having type-tied names is cool, but type `declare`ations would work better. That's what I remember from the talk on my habit to put type requirements in docstrings.
- While `quri:uri` has it's name, we're mainly using it to store the website links, URLs. A meaningful name would be URL in this situation, while URI is the format, not meaning, and type delivers the format idea it well enough for it be in the name.

**Comment by jmercouris** [comment]
Artyom has made a very good point here, I'm starting to doubt my earlier position :-D. There are however inconsistencies here, we use `-p` for example to signify that a symbol is pointing to a predicate. Type declarations are great, but it is also nice to have a symbol that makes things clear without having to reference them. What about this?:

URI when using URI
and URL-string when using a string representation

**Comment by aartaka** [comment]
Hmmm, given that string representation is an exception now, it's sounds good to make it stand out. But, outside this, we use `quri:uri` to store locations on the web by default and in most cases. I think this default behavior doesn't need explicit signs outside type declarations and documentation.

**Comment by aartaka** [comment]
So I'm still on the side of `url` name for URLs, no matter what their format is, and `-string` postfix for the string-represented ones.

**Comment by Ambrevar** [comment]
We might be bike-shedding here a little :)

**Comment by aartaka** [comment]
So, what stops the merging?

**Comment by Ambrevar** [comment]
Nothing, I just need to find time to test and merge today.  Will do, no worries!

**Comment by Ambrevar** [comment]
This is a great change, thanks a lot for your awesome contribution again!

**Comment by jmercouris** [comment]
+1 Well done!
